File: 121.java
import java.util.Arrays;

class Solution121 {
    public int maxProfit(int[] prices) {
        int max_profit = 0;
        int[] maxes  = new int[prices.length + 1];
        maxes[prices.length] = 0;
        for (int i = prices.length -1; i > 0; i--){
            maxes[i] = Math.max(prices[i], maxes[i + 1]);
        }
        for (int i = 0; i < prices.length; i++){
            if (maxes[i] - prices[i] > max_profit){
                max_profit = maxes[i] - prices[i];
            }
        }
        return max_profit;
    }
}

File: 121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy_price = prices[0]
        profit = 0
        for price in prices:
            if price < buy_price:
                buy_price = price
            profit = max(profit, price - buy_price)
        return profit


s1 = Solution()
p = s1.maxProfit([7,6,4,3,1])
print(p)